src/game_logic/test.rs

```diff
@@ -0,0 +1,566 @@
+    assert!(matches!(
+        game.players[0].borrow().block_or_timer,
+        BlockOrTimer::TimerPending
+    ));
+    // The not-fitting block disappeared. Not super important, feel free to change.
+    assert_eq!(
+        dump_game_state(&game),
+        [
+            "                    ",
+            "            LL      ",
+            "        LLLLLL      ",
+        ]
+    );
+
+    // We can now query whose timers are pending, but we get None to indicate game over.
+    assert!(game.start_pending_please_wait_counters().is_none());
+}
+
+#[test]
+fn test_wait_counters() {
+    let mut game = create_game(Mode::Traditional, 2);
+    game.truncate_height(3);
+
+    game.move_blocks_down(false);
+    game.move_blocks_down(false);
+    game.move_blocks_down(false);
+    game.move_blocks_down(false);
+    game.move_blocks_down(false);
+    assert_eq!(
+        dump_game_state(&game),
+        [
+            "    FFFFFF        FFFFFF    ",
+            "            LL        LL    ",
+            "        LLLLLL    LLLLLL    ",
+        ]
+    );
+    assert_eq!(game.start_pending_please_wait_counters(), Some(vec![]));
+
+    // Player 0 (left) can still keep going, but player 1 (right) starts their 30sec waiting time
+    game.move_blocks_down(false);
+    assert_eq!(
+        dump_game_state(&game),
+        [
+            "        FF                  ",
+            "    FFFFFF  LL        LL    ",
+            "        LLLLLL    LLLLLL    ",
+        ]
+    );
+    assert!(matches!(
+        game.players[1].borrow().block_or_timer,
+        BlockOrTimer::TimerPending
+    ));
+    assert_eq!(game.start_pending_please_wait_counters(), Some(vec![1]));
+    assert!(matches!(
+        game.players[1].borrow().block_or_timer,
+        BlockOrTimer::Timer(30)
+    ));
+
+    // During the next 30 seconds, the timer ticks from 30 to 1. Then the player gets a new block.
+    for _ in 0..28 {
+        assert!(game.tick_please_wait_counter(1));
+    }
+    assert!(matches!(
+        game.players[1].borrow().block_or_timer,
+        BlockOrTimer::Timer(2)
+    ));
+    assert!(game.tick_please_wait_counter(1));
+    assert!(matches!(
+        game.players[1].borrow().block_or_timer,
+        BlockOrTimer::Timer(1)
+    ));
+    assert!(!game.tick_please_wait_counter(1));
+    assert!(matches!(
+        game.players[1].borrow().block_or_timer,
+        BlockOrTimer::Block(_)
+    ));
+}
+
+#[test]
+fn test_traditional_clearing() {
+    let mut game = create_game(Mode::Traditional, 2);
+    game.truncate_height(5);
+    for y in 1..5 {
+        for x in 0..(game.get_width() as i16) {
+            if (x, y) != (5, 2) && (x, y) != (8, 4) && (x, y) != (12, 4) {
+                game.set_landed_square(
+                    (x, y),
+                    Some(SquareContent::with_color(Color::YELLOW_FOREGROUND)),
+                );
+            }
+        }
+    }
+    let before_clear = vec![
+        "                            ",
```

Comment:
Should probably add something to this top row too

Reply:
Not really necessary. Non-cleared between two cleareds checks it from both sides.